### PR TITLE
Workflow exports links

### LINKED ISSUE #3693: Make exports by workflow public
We have three launched approved projects now using exports by workflow without any bug reports. It should be made public.

### PR DESCRIPTION
**Pull request by jelliotartz**
Fixes #3693.

Describe your changes.
- Replace select box with radio button list.
- API request to fetch export media.
- Links to previous exports.
- Only export every 24 hours.

https://workflow-exports-links.pfe-preview.zooniverse.org/lab/1663/data-exports

# Review Checklist

- [x] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [x] Does it work on mobile?
- [x] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [x] Did you deploy a staging branch?
- [x] Did you get any type checking errors from [Babel Typecheck](https://github.com/codemix/babel-plugin-typecheck)

## Optional

- [x] If it's a new component, is it in ES6? Is it clear of warnings from ESLint?
- [x] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [x] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?

### CODE REVIEW

File: app/pages/lab/export-workflows-dialog.jsx [hunk 1]
```diff
@@ -36,17 +42,25 @@ class ExportWorkflowsDialog extends React.Component {
...
+        // TODO: load the exports for each workflow
+        workflows.forEach((wf) => {
+          wf.get('classifications_export')
+            .then((media) => {
+              const mediaState = this.state.media;
+              mediaState[wf.id.toString()] = media[0];
+              this.setState({ media: mediaState });
+            })
+            .catch(() => {
+              // this is gonna happen, oh well
```
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: Is there going to be a handler here and any sort of error messaging to the user for failure?

File: app/pages/lab/export-workflows-dialog.jsx [hunk 2]
```diff
@@ -36,17 +42,25 @@ class ExportWorkflowsDialog extends React.Component {
...
         this.setState({ workflows, loading: false });
-      });
-  }
 
-  toggleExport() {
-    if (this.workflowList.selectedIndex >= 0) {
-      this.setState({ workflowSelected: true });
-    }
+        // TODO: load the exports for each workflow
+        workflows.forEach((wf) => {
```
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: For consistency and clarity, I'd prefer `workflow` instead of `wf`

File: css/workflow-export-dialog.styl [hunk 3]
```diff
@@ -0,0 +1,11 @@
+.workflow-export-list
+  padding-left: 0
+
+  li
```
Comment (srallen) [inline css/workflow-export-dialog.styl]: Let's actually target these with a class rather than the tag. We've been working toward using a BEM convention for class names.

File: app/pages/lab/export-workflows-dialog.jsx [hunk 4]
```diff
@@ -90,7 +105,50 @@ class ExportWorkflowsDialog extends React.Component {
...
+      <small>
+        <ExportWorkflowLink media={myMedia} />
+      </small>
+    </li>
+  );
+};
+
+ExportWorkflowListItem.propTypes = {
+  workflow: React.PropTypes.shape({ id: React.PropTypes.string, display_name: React.PropTypes.string }).isRequired,
+  media: React.PropTypes.shape({}).isRequired,
```
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: This seems to be missing a shape definition.

File: app/pages/lab/export-workflows-dialog.jsx [hunk 5]
```diff
@@ -90,7 +105,50 @@ class ExportWorkflowsDialog extends React.Component {
...
+      <input type="radio" title={titleString} name="which-workflow" id={`export-${workflow.id}`} disabled={lockout} onChange={onChange} />
+      {workflow.display_name}
+      <small>
+        <ExportWorkflowLink media={myMedia} />
+      </small>
+    </li>
+  );
+};
+
+ExportWorkflowListItem.propTypes = {
```
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: The props that are set as required should have defined defaults.

File: app/pages/lab/export-workflows-dialog.jsx [hunk 6]
```diff
@@ -90,7 +105,50 @@ class ExportWorkflowsDialog extends React.Component {
 ExportWorkflowsDialog.propTypes = {
```
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: The required prop types should have defaults.

File: app/pages/lab/export-workflows-dialog.jsx [hunk 7]
```diff
@@ -90,7 +105,50 @@ class ExportWorkflowsDialog extends React.Component {
...
   project: React.PropTypes.shape({ links: React.PropTypes.object }).isRequired,
   onSuccess: React.PropTypes.func.isRequired,
-  onFail: React.PropTypes.func.isRequired,
+  onFail: React.PropTypes.func.isRequired
+};
+
+const ExportWorkflowListItem = ({ workflow, media, onChange }) => {
+  const myMedia = workflow ? media[workflow.id.toString()] : null;
```
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: The workflow id is already a string. This conversion does not need to happen.

File: app/pages/lab/export-workflows-dialog.jsx [hunk 8]
```diff
@@ -36,17 +42,25 @@ class ExportWorkflowsDialog extends React.Component {
...
-  toggleExport() {
-    if (this.workflowList.selectedIndex >= 0) {
-      this.setState({ workflowSelected: true });
-    }
+        // TODO: load the exports for each workflow
+        workflows.forEach((wf) => {
+          wf.get('classifications_export')
+            .then((media) => {
+              const mediaState = this.state.media;
+              mediaState[wf.id.toString()] = media[0];
```
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: Workflow id is already a string. `toString()` is not necessary.

File: app/pages/lab/export-workflows-dialog.jsx (hunk 1 above)
Comment (amyrebecca) [inline app/pages/lab/export-workflows-dialog.jsx]: This will fail with a 404 for any workflow that hasn't been exported yet. I'm not crazy about that design decision because it makes us swallow this error. It doesn't affect the user experience at all, which is why we don't do any messaging.

File: app/pages/lab/export-workflows-dialog.jsx (hunk 4 above)
Comment (amyrebecca) [inline app/pages/lab/export-workflows-dialog.jsx]: We don't use any parts of the shape in this component, because we just pass it on to the child component. The child component validates the props that it cares about, so it would probably be redundant to do it here. The linter rules don't let us simply specify that it's an object and is required, so this is the best solution we could think of.

File: css/workflow-export-dialog.styl (hunk 3 above)
Comment (amyrebecca) [inline css/workflow-export-dialog.styl]: I feel weird about adding styles like list-style-type to classes instead of an `li` tag. We did intend to use BEM styles for this component, but it didn't seem necessary because we were able to use styling on the semantic elements. Does BEM really prefer classes to tags?

**Comment by amyrebecca** [comment]
I really dislike making separate files for components like ExportWorkflowListItem that don't make sense outside the context of a parent component--it just seems like it clutters the filesystem for no good reason. But if other folks feel strongly the opposite way I'm not going to be dogmatic about it obviously :)

File: app/pages/lab/export-workflows-dialog.jsx (hunk 1 above)
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: We fixed the error handling in the json-api-client so you can appropriately handle different error codes. So this could be something like `if (error.status !== 404) doSomething`. See example with the classifier: https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/pages/project/classify.cjsx#L281-L288

File: app/pages/lab/export-workflows-dialog.jsx (hunk 4 above)
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: I'd rather have it defined as an object then and add an inline linter rule exception rather than define it as a shape since that just looks like a mistake.

File: css/workflow-export-dialog.styl (hunk 3 above)
Comment (srallen) [inline css/workflow-export-dialog.styl]: Yes: http://getbem.com/faq/#custom-tags-for-blocks

**Comment by srallen** [comment]
Separate files for each component is the Airbnb style that I think we should follow, but the rule was disabled because a lot of our old coffeescript files were written that way. IMO, it really does make it hard to follow what component is doing what, especially if you're trying to debug something you didn't write, and bloats individual files. I'd rather have the bloat in the filesystem as it's much more clear.

File: app/pages/lab/export-workflows-dialog.jsx (hunk 1 above)
Comment (amyrebecca) [inline app/pages/lab/export-workflows-dialog.jsx]: Oh that's handy! Still, even if the error were anything other than 404, I'm not sure we'd want to message the user about it, given that we're not performing the API request in response to a user interaction, just doing it automatically.

File: app/pages/lab/export-workflows-dialog.jsx (hunk 4 above)
Comment (amyrebecca) [inline app/pages/lab/export-workflows-dialog.jsx]: Works for me, I've done that in the past too.

File: css/workflow-export-dialog.styl (hunk 3 above)
Comment (amyrebecca) [inline css/workflow-export-dialog.styl]: Sounds good, I'll do that from here on out.

File: app/pages/lab/export-workflows-dialog.jsx (hunk 1 above)
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: Yes, I see what you mean, so perhaps `if (error.status === 404) return`? FWIW, in a future iteration, all of the export API calls could be refactored to use an include on the project.

Comment (amyrebecca) [inline app/pages/lab/export-workflows-dialog.jsx]: That's a good idea--even though that's what would happen by default anyways, it's a bit more self-documenting.

And you're right, it might also be nice to take a little load off the API to get one request (for the project with its associated export media) instead of five.

File: app/pages/lab/export-workflows-dialog.jsx (hunk 7 above)
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: @jelliotartz It looks like this comment still needs to be addressed.

File: app/pages/lab/export-workflows-dialog.jsx [hunk 9]
```diff
@@ -36,17 +42,28 @@ class ExportWorkflowsDialog extends React.Component {
...
-    }
+        workflows.forEach((workflow) => {
+          workflow.get('classifications_export')
+            .then((media) => {
+              const mediaState = this.state.media;
+              mediaState[workflow.id] = media[0];
+              this.setState({ media: mediaState });
+            })
+            .catch((error) => {
+              if (error.status === 404) {
```
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: I think this is the reverse of what we were chatting about. The 404 just indicates the project owner hasn't requested this export before, so we don't really care about reporting that back to the user. What I think this should be is:
```
if (error.status !== 404) {
  this.props.onFail.(error);
}
```

**Comment by srallen** [comment]
I found that when the export GET requests were randomly returning a 500 error, that the error message is showing up in the UI on the page rather than in the modal. I think a generic error message should be displayed in the modal for the GET request, maybe inline with the particular workflow export that failed? In the below image, the Journey workflow is the GET request that erred. 

screen shot 2017-05-31 at 9 48 34 am

**Comment by jelliotartz** [comment]
I'm having trouble reproducing the 500 error - I've tried emptying cache and hard reload in Chrome about 10 times, and all I'm seeing in the console are 404 errors from the GET request (and no error message in the UI on the page). Am I missing something? or is there another way I can produce the 500 error besides empty cache & hard reload && wait for random occurrence?

**Comment by jelliotartz** [comment]
I passed `exportError` from `WorkflowExportClassificationButton`'s state to `ExportWorkflowsDialog` as a prop, but its currently displaying `exportError` beneath the list items, rather than inline - I realize this isn't what you're asking for, but I'm at a loss to figure out how to add `exportError` to an individual `ExportWorkflowListItem` (and display it inline, beneath it). 

When I pass `exportError` as a prop to `ExportWorkflowListItem`, it displays it beneath every item. Do I need to find a way to append `exportError` to an individual item? since it starts as part of `WorkflowExportClassificationButton`'s state, it seems like it therefore belongs to all its children if I pass it to them... I could use a little help figuring out what I'm missing to get the inline display with a failed workflow.

**Comment by srallen** [comment]
I think you're on the right track. Instead of trying to reuse the error handler from the parent, you could add a handler in the `ExportWorkflowDialog` component class and a state property. If a particular workflow errors, then update the error state in the catch of the promise for the get request maybe like:
```
let workflowError = {};
workflowError[workflow.id] = error;
this.setState({ workflowError })
```
(I'm using the [object property shorthand syntax](http://eslint.org/docs/rules/object-shorthand) in setState in the above example)

Then pass down that error state as a prop to the list item components and determine which error belongs to which workflow using the id. I'd probably rework those two components into one and instead of using a ternary, have a few inline conditionals (examples here: https://facebook.github.io/react/docs/conditional-rendering.html#inline-if-with-logical--operator) for the various rendering cases.

**Comment by amyrebecca** [comment]
Would it be cleaner to simply leave the error where it is and dismiss the modal after the "export" button is clicked? If not, then the stuff you've outlined would be an ideal solution, I just thought I'd suggest something simple in case it worked :)

**Comment by srallen** [comment]
I think for the export POST, yes, that's how the error handling should work, but for the initial GET requests when the modal opens, I think showing inline which workflow GET request failed would be a better UI. In my example, if the request for the Journey workflow fails like it did when I tried it, then I shouldn't be stopped from being able to click to download the export from the Master workflow and should also be able to clearly understand that the Journey workflow link is failing.

**Comment by jelliotartz** [comment]
I'm having trouble testing whether this displays errors inline with the correct workflow. In addition to following @srallen's suggestions, I used the setState of media as a guide, and because this is an object, I'm not getting the ability to update `workflowError` in the React dev tools console like I could when `exportError` was initially set as an empty string. Is there a way I can test whether this is handling and displaying errors correctly, if the prop isn't a string? Or should I be approaching this differently?

Also, happy to refactor `ListItem` and `Dialog` into one component once the error is handled and displayed correctly, just wanted to get that sorted out first :)

File: app/pages/lab/export-workflows-dialog.jsx [hunk 10]
```diff
@@ -36,17 +43,30 @@ class ExportWorkflowsDialog extends React.Component {
...
+            .then((media) => {
+              const mediaState = this.state.media;
+              mediaState[workflow.id] = media[0];
+              this.setState({ media: mediaState });
+            })
+            .catch((error) => {
+              if (error.status !== 404) {
+                this.props.onFail(error);
+              } else {
+                const workflowErrorState = this.state.workflowError;
```
Comment (srallen) [inline app/pages/lab/export-workflows-dialog.jsx]: This is the error handling that should be done instead of `onFail` for the get request for errors that are not 404. This should be moved into the true block of the if statement.
